lib/Smr/Container/DiContainer.php

```diff
@@ -0,0 +1,112 @@
+<?php declare(strict_types=1);
+
+namespace Smr\Container;
+
+use DI\Container;
+use DI\ContainerBuilder;
+use Dotenv\Dotenv;
+use MySqlDatabase;
+use mysqli;
+use RuntimeException;
+use Smr\MySqlProperties;
+use function DI\autowire;
+
+/**
+ * Class DiContainer
+ * A wrapper around the DI\Container functionality that will allow
+ * static usage of container methods.
+ * @package Smr\Container
+ */
+class DiContainer {
+	private static DiContainer $instance;
+	private Container $container;
+
+	private function __construct() {
+		self::$instance = $this;
+		$this->container = $this->buildContainer();
+	}
+
+	private function getDefinitions(): array {
+		return [
+			/*
+			 * mysqli is a 3rd party library, and we do not have control over its constructor.
+			 * In order for PHP-DI to construct an instance of a class, each constructor argument must
+			 * be able to be constructed.
+			 * PHP-DI cannot construct mysqli by itself, because all of its arguments are primitive types.
+			 * Therefore, we need to declare a provider factory for the container to use when constructing new instances.
+			 *
+			 * The factories themselves are able to use dependency injection as well, so we can provide the MysqlProperties
+			 * typehint to make sure the container constructs and instance and provides it to the factory.
+			 */
+			mysqli::class => function (MySqlProperties $mysqlProperties): mysqli {
```

Comment:
Can this anonymous function be moved to a static function of `MySqlDatabase`? I think that if a developer wants to understand how we interact with the database, they should only really need to look at the `MySqlDatabase` class. The `DiContainer` class is probably the last place I'd expect to find it. :)

Does that seem reasonable, or is there a good argument for keeping it here?

Reply:
All it's doing is returning a `mysqli` object -- all the other interactions are already happening in the `MySqlDatabase` class. This is a factory method for producing the `mysqli` dependency required by the database class to operate. I'd be willing to add some more documentation around the constructor for the `MySqlDatabase` class, but I think this is the appropriate place for this dependency factory to be. It's not immediately obvious in this iteration, but there is a purpose for this being here :D

Reply:
It's a dependency of the class, so why should it not be injected as a dependency? It makes it near impossible to do simple tests otherwise. If I want to test failure scenarios, or business logic on that dependency it cannot be done easily, or concisely.

I feel like you can clearly see how this class is using the `mysqli` dependency, and I don't know how having it's construction right in this class helps understanding anything extra; there's only one way to construct that thing :D

New developers who don't understand about dependency injection techniques can just ask questions, or we can have more documentation in the readme about the paradigm shift happening in the codebase.

I also feel like "I don't understand it" is not a valid reason to skimp out on modern best practices for software architecture, but if you feel so strongly about this then I'm willing to concede the whole thing.

However, I can't guarantee you the test coverage the codebase deserves without using this tried-and-true pattern.

Reply:
I'm not advocating for a change in the dependency injection scheme, just moving the guts of the factory method to a static function of `MySqlDatabase` so that it's the one-stop-shop for seeing how our class wraps the `mysqli` library.

Let me lay out a sketch of what I'd like to see. For example, in `DiContainer.php`, we'd have:
```php
mysqli::class => function (MySqlProperties $mysqlProperties): mysqli {
    return MySqlDatabase::createMysqliObject($mysqlProperties);
}
```
And in `MySqlDatabase.class.php`, we'd have:
```php
/**
 * This method will be used by the DI factory to create the mysqli object, etc.
 */
public static function createMySqliObject(MySqlProperties $mysqlProperties): mysqli {
    // Set the mysqli driver to raise exceptions on errors
    if (!mysqli_report(MYSQLI_REPORT_ERROR | MYSQLI_REPORT_STRICT)) {
        throw new RuntimeException('Failed to enable mysqli error reporting');
    }
    return new mysqli(
        $mysqlProperties->getHost(),
        $mysqlProperties->getUser(),
        $mysqlProperties->getPassword(),
        $mysqlProperties->getDatabaseName());
    // And if I were writing this, I'd put the utf-8 sanity check here so that it is performed once per connection. YMMV!
}
```

Functionally, I don't think this is any different than what you're doing. Organizationally, you may still dislike it, but I believe it should still allow you to perform all the testing you want to do.

Does this clarification change the equation at all for you?

Reply:
Yeah, it's a compromise I can live with. But now you have one of those pesky `public static` methods hanging around

Reply:
Moved